### ISSUE TITLE
segmentation fault on require without Gtk.main()

### ISSUE THREAD

**Issue by ewnd9** [Place1/node-gir#15]
This works:

```js
const { Gtk } = require('../');
Gtk.main();
```

This throws segmentation fault:

```js
const { Gtk } = require('../');
```

**Comment by Place1** [comment]
The following also seg faults.

```
const { Gtk } = require('./');
setTimeout(() => Gtk.mainQuit(), 100);
Gtk.main();
```

The issue is related to some cleanup code within the bindings. I'm still investigating.

**Comment by jamrizzi** [comment]
Yeah, I've been exiting gtk with process.exit() because of the segfault.

**Comment by jamrizzi** [comment]
Where is the cleanup code?

**Comment by Place1** [comment]
It crashes when the map in struct.cpp is destroyed I believe. But I’m not
sure.

**Comment by jamrizzi** [comment]
Ok, thanks

**Comment by cedlemo** [comment]
any updates on this issue ?

**Comment by Place1** [comment]
We have an idea on where it is but I haven’t had time to try and fix it yet
unfortuantly.

The bug is obviously bad but it only occurs when the process exits, so it
shouldn’t get in your way playing around with the library.

**Comment by cedlemo** [comment]
> We have an idea on where it is but I haven’t had time to try and fix it yet
unfortuantly.

If you have any idea I can work on to try to fix it please let me know.

> The bug is obviously bad but it only occurs when the process exits, so it
shouldn’t get in your way playing around with the library.

Ok. I will try to do some samples from https://github.com/GNOME/gtk/tree/master/examples and eventually send some PR or open some issues if I have some difficulties to adapt them.

**Comment by Place1** [comment]
@cedlemo i think the issue is related to the [GIRStruct::prepared_js_classes](https://github.com/Place1/node-gir/blob/e3ce1dfe9579ce530d765122d5b5aada8391c68e/src/types/struct.cpp#L19) map being deallocated. Perhaps we need a custom destructor to handle cleaning up `Nan::Persistent<FunctionTemplate, CopyablePersistentTraits<FunctionTemplate>>` instances rather than the default destructor? Not really sure though. Getting a better idea on the cause (probably via `gdb` or `valgrind`) might be useful.

**Comment by backwards-rat-race** [comment]
Hi, don't know if there's been any updates on this but I had a quick look based on what @Place1 about the de-allocation and created a really quick and dirty wrapper around that object. The wrapper just calls .Empty() on the each of the NanPersistant objects in the map during the destructing and it appears to fix the Segfault issue.

https://github.com/backwards-rat-race/node-gir/commit/d607baba96ad55bc348c19a619913f72bbc0c4f4

I plan on creating a proper wrapper around the object soon and I can make a pull request, but didn't know if that was the best way of solving the issue.

**Comment by Place1** [comment]
@backwards-rat-race this is interesting. What's the difference between `.Empty()` and `.Reset()` for a Nan:Persistent?

https://github.com/nodejs/nan/blob/master/doc/persistent.md

I couldn't find any docs on it :(

**Comment by backwards-rat-race** [comment]
Looking at the Nan code and V8 docs, `.Reset()` clears the handles reference and then sets the handle to 'empty' so any subsequent calls to `isEmpty()` will return true. `.Empty()` will just clear the handle, which means `isEmpty()` will only return true if the GC/Runtime has cleaned up the handle.

I believe `.Reset()` is designed to be used for when you want to change the handle that Nan:Persistent  is looking at. Whereas `.Empty()` is more to say this memory is done with, which is the case for us here.

[`Reset`](https://github.com/nodejs/nan/blob/77d0fcaba3305d05176a9ad95d8e5101e8f2a283/nan_persistent_pre_12_inl.h#L27) calls `.dispose()` and `.clear()`, but [`Empty`](https://github.com/nodejs/nan/blob/77d0fcaba3305d05176a9ad95d8e5101e8f2a283/nan_persistent_pre_12_inl.h#L64) just calls `.dispose()`. [`Reset`](https://github.com/nodejs/nan/blob/77d0fcaba3305d05176a9ad95d8e5101e8f2a283/nan_persistent_pre_12_inl.h#L33) also has the ability to swap out the handle with a new reference.

**Comment by Place1** [comment]
@backwards-rat-race i'd be happy to merge your PR. It's a simple enough fix :D

**Comment by backwards-rat-race** [comment]
I ended up debugging the V8 persistence object a bit more and found a more 'correct' solution which fixes the issue rather than adding a wrapper which seems a bit hack-y.
I've requested a pull request with the code changes, I don't use github much so hope it's all alright. Thanks.
